server/s3_delete_object_action.cc

```diff
@@ -125,6 +134,91 @@ void S3DeleteObjectAction::fetch_object_info_failed() {
   s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
 }
 
+void S3DeleteObjectAction::delete_handler() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  std::string bucket_versioning =
+      bucket_metadata->get_bucket_versioning_status();
+
+  if (bucket_versioning == "Enabled" &&
+      !request->has_query_param_key("versionId")) {
+    create_delete_marker();
+    next();
+  } else {
+    populate_probable_dead_oid_list();
+  }
+
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3DeleteObjectAction::metadata_handler() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  std::string bucket_versioning =
+      bucket_metadata->get_bucket_versioning_status();
+  if (bucket_versioning == "Enabled" &&
+      !request->has_query_param_key("versionId")) {
+    save_delete_marker();
+  } else {
+    delete_metadata();
+  }
+
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3DeleteObjectAction::create_delete_marker() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  // create metadata
+  delete_marker_metadata =
+      object_metadata_factory->create_delete_marker_metadata_obj(
+          request, bucket_metadata->get_object_list_index_layout(),
+          bucket_metadata->get_objects_version_list_index_layout());
+
+  // Generate a version id for the new object.
+  delete_marker_metadata->regenerate_version_id();
+  delete_marker_metadata->reset_date_time_to_current();
+
+  // TODO: update null version
+
+  s3_log(S3_LOG_DEBUG, request_id, "Add delete marker4\n");
+  for (auto it : request->get_in_headers_copy()) {
+    if (it.first.find("x-amz-meta-") != std::string::npos) {
```

Comment:
Shouldn't it be:

```
    if (it.first.find("x-amz-meta-") == 0) {
```

Otherwise, it will find `x-amz-meta` anywhere in the string.

Reply:
Yes i was unaware of tis part
after testing
updated code